x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportCreateAndFollowIndexAction.java

```diff
@@ -224,19 +211,51 @@ public ClusterState execute(final ClusterState currentState) throws Exception {
+                IndexMetaData followerIndexMeta = newState.metaData().index(request.getFollowRequest().getFollowerIndex());
+                initiateFollowing(request, followerIndexMeta, listener);
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
         });
     }
 
     private void initiateFollowing(
             final CreateAndFollowIndexAction.Request request,
+            final IndexMetaData followerIndexMeta,
             final ActionListener<CreateAndFollowIndexAction.Response> listener) {
         activeShardsObserver.waitForActiveShards(new String[]{request.getFollowRequest().getFollowerIndex()},
                 ActiveShardCount.DEFAULT, request.timeout(), result -> {
                     if (result) {
-                        client.execute(FollowIndexAction.INSTANCE, request.getFollowRequest(), ActionListener.wrap(
-                                r -> listener.onResponse(new CreateAndFollowIndexAction.Response(true, true, r.isAcknowledged())),
-                                listener::onFailure
-                        ));
+                        ccrLicenseChecker.fetchHistoryUUIDs(client, followerIndexMeta, listener::onFailure, historyUUIDs -> {
+                            clusterService.submitStateUpdateTask("", new ClusterStateUpdateTask() {
+                                @Override
+                                public ClusterState execute(ClusterState currentState) throws Exception {
+                                    return recordFollowerShardHistoryUUIDs(currentState, followerIndexMeta.getIndex(), historyUUIDs);
+                                }
+
+                                @Override
+                                public void onFailure(String source, Exception e) {
+                                    listener.onFailure(e);
+                                }
+
+                                @Override
+                                public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                                    client.execute(FollowIndexAction.INSTANCE, request.getFollowRequest(), ActionListener.wrap(
+                                        r -> listener.onResponse(new CreateAndFollowIndexAction.Response(true, true, r.isAcknowledged())),
+                                        e -> {
+                                            logger.error("Failed to initiate the shard following", e);
+                                            listener.onResponse(new CreateAndFollowIndexAction.Response(true, true, false));
+                                        }
+                                    ));
+                                }
+                            });
+                        });
                     } else {
                         listener.onResponse(new CreateAndFollowIndexAction.Response(true, false, false));
```

Comment:
I wondering if we should throw an exception here. Normally when something isn't acked, if you wait it will go through (or at has a chance of doing it). This one will just get stuck.